test/test_image.py

```diff
@@ -134,20 +168,26 @@ def test_write_jpeg(self):
             self.assertEqual(torch_bytes, pil_bytes)
 
     def test_decode_png(self):
+        conversion = [(None, 0), ("L", 1), ("LA", 2), ("RGB", 3), ("RGBA", 4)]
         for img_path in get_images(FAKEDATA_DIR, ".png"):
-            img_pil = torch.from_numpy(np.array(Image.open(img_path)))
-            if len(img_pil.shape) == 3:
-                img_pil = img_pil.permute(2, 0, 1)
-            else:
-                img_pil = img_pil.unsqueeze(0)
-            data = read_file(img_path)
-            img_lpng = decode_png(data)
-            self.assertTrue(img_lpng.equal(img_pil))
+            for pil_mode, channels in conversion:
+                with Image.open(img_path) as img:
+                    if not (pil_mode == "L" and img.mode == "P") and pil_mode is not None:
```

Comment:
My current implementation does not convert palette images to grayscale. This is not a bug/limitation but a feature™. This is because palette images, which can be useful for image segmentation, already have only 1 channel. So when a user specifies `channels=1` there is no need do an additional conversion to grayscale.

Reply:
I agree that this complicates the semantics. On the other hand, it does not have a real effect on the situation once it becomes a tensor. 

At any case, I do not mind changing this but it would probably involve a double conversion first to RGB and then to Grayscale because jpeglib does not support a Palette to Grayscale conversion. Thoughts?

Reply:
There is a semantics change in converting it or not to grayscale for the following reason:
some datasets store semantic maps as a palette image, with labels / colors going from 0 to 20 or so. If we were expecting to see the image in `L` format, we would see it with different colors.

Here is one example:
```python
i = Image.open('/datasets01/VOC2007/011719/SegmentationClass/005130.png')
print(i.mode)
# > 'P'

# color values used as labels for semantic segmentation
np.unique(np.array(i))
# > array([  0,   3, 255], dtype=uint8)

# but if we convert them to grayscale for visualization
# values are different
np.unique(np.array(i.convert('L')))
# > array([  0, 113, 220], dtype=uint8)
```

> double conversion first to RGB and then to Grayscale

That's fine with me

Reply:
Sounds good I'll make the change.